### DISCUSSION TITLE
move special treatment to a user specific file

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#16] [Ideas]
The kodi_panel_display.py script contain various "special treatment" eg: artist, codec, acodec, full_codec etc
In My case I have changed some of them to get different behaviour (eg: different concatenation on full_codec)
Having a separate user file to define theses treatment will allow to update main script without loosing the treatments.

**Comment by mattblovell** [comment]
Capturing some notes on possible implementation:

- Special treatments could all be consolidated to a dictionary, with each dictionary key corresponding to the "special" `name` for a text field, and the entry value a function to invoke to obtain the string to display.  This convention would require a uniform signature for all those callback functions.

- An invoking script would then be free to modify either the function for already-existing entries or to add new entries.  Overriding the existing functions would be much like what occurs in for the `screen_on()` and `screen_off()` overrides for PWM within the current `kodi_panel_fb.py`.

- The list of keys in this "special treatment" dictionary could permit for the additional string interpolation proposed in Idea #17.

**Comment by mattblovell** [comment]
Opened enhancement issue #21 for the above.

**Comment by mattblovell** [comment]
Should be possible with v1.22

**Comment by mattblovell** [comment]
> Should be possible with v1.22

Expanding on the above, whatever script the user actually invokes (e.g., kodi_panel_fb.py) has the ability, prior to invoking `kodi_panel_display.main()`, to modify the two lookup tables -- `ELEMENT_CB` and `STRING_CB`.

New entries can be added to either dictionary, existing entries can be modified (pointing to a user-specific function), and existing entries can be deleted.

The user's setup.toml file can specify layout elements / fields that make use of those new "special treatments" by using the same string as gets specified for the dictionary entries.

**Comment by nico1080** [comment]
To be sure to understand:

If I modify in my `kodi_panel_fb.py`:

define a function in my `kodi_panel_fb.py`
```
def strcb_vcodec(info, screen_mode, layout_name):
    if 'VideoPlayer.AudioCodec' in info:
        if info['VideoPlayer.VideoCodec'] in codec_name.keys():
            return codec_name[info['VideoPlayer.VideoCodec']]
        else:
            return info['VideoPlayer.VideoCodec']
    return ""
```
And then add it to the `STRING_CB` lookup table?
`kodi_panel_display.STRING_CB["vcodec"] = strcb_vcodec`
(I am not sure this syntax would add it? )

I could then invoke a vcodec field in any of my layouts?

Alternatively I could use any heuristic with screen_mode and/or layout_name to change the returned value

**Comment by mattblovell** [comment]
> I could then invoke a vcodec field in any of my layouts?

Yes, that's the idea.

Your syntax for adding the new entry to the `STRING_CB` table looks to be correct.  The proposed `strcb_vcodec()` function just has to be defined in your kodi_panel_fb.py file someplace ahead of the line where you make that assignment.  (Python does not have what are termed "forward references".)

The only other caveat regards the `codec_name[]` table lookup.  Since that table is defined within kodi_panel_display.py, I think the namespace has to be explicitly specified.  The lookup in the return statement would thus be

`return kodi_panel_display.codec_name[info['VideoPlayer.VideoCodec']]`

That somewhat lengthy statement _can_ be shortened, I believe, if one changes the `import` statement earlier within kodi_panel_fb.py.  Using something like

`import kodi_panel_display as kp`

The `as` effectively lets you rename the module as you import it.  The return statement could then be shortened to:

`return kp.codec_name[info['VideoPlayer.VideoCodec']]`

Python documentation on import is available online:

https://docs.python.org/3/reference/simple_stmts.html#import

Similar namespace-qualification is likely needed anywhere you need to refer to global variables or types that are defined within the kodi_panel_display.py file.  I think that's really the primary hurdle for augmenting kodi_panel functionality via "external" code.

(Python also has an `from <module> import *` mechanism that could serve to eliminate the need for namespace qualification.  To take advantage of that mechanism, I think I need to define an `__all__` variable that specifies what all the _public names_ are for kodi_panel_display.)

**Comment by nico1080** [comment]
Very clear on that, I think I will create a new branch on my fork that will just copy kodi_panel_display.py from your master and make all my special treatment in the starting script.

One other though, would it not be better to move the `codec_name[]` table in the start script and / or `setup.toml` (after all it is user value)
In my case, I think I will use `custom_codec_name table[]` for all my personnal script

**Comment by mattblovell** [comment]
> One other though, would it not be better to move the codec_name[] table in the start script and / or setup.toml

If the table / dictionary resided in the startup script, then code within kodi_panel_display.py wouldn't have any access to it.  (As kodi_panel_display is what gets imported, it doesn't get to "know" the name of the importing module.)

Giving `setup.toml` the ability to _augment_ the table (with the freedom to overwrite existing entries) is certainly possible.  I can look into that.

**Comment by mattblovell** [comment]
The to-be-released v1.30 permits the `codec_name` dictionary to be updated from the setup.toml file.

**Comment by nico1080** [comment]
I confirm every thing work

I have added a lookup for video codec as mentionned: see the commit in my fork https://github.com/nico1080/kodi_panel/commit/0ec9fda57acfb71e715e0a18746af49a94db5554

And also a lookup for channel number that will send back `5.1` instead `6 `https://github.com/nico1080/kodi_panel/commit/bc6e2e4365f6fc00d87f5c8f1076cb0452ad2c6c
